python/lsst/ctrl/mpexec/cli/cmd/commands.py

```diff
@@ -126,44 +121,40 @@ def build(ctx: click.Context, **kwargs: Any) -> None:
     _unhandledShow(show, "build")
 
 
-def _start_coverage(coverage_packages: tuple) -> coverage.Coverage:
-    coveragerc = """
-[html]
-directory = covhtml
-
+@contextmanager
+def coverage_context(kwargs: dict[str, Any]) -> Iterator[None]:
+    packages = kwargs.pop("cov_packages", ())
+    report = kwargs.pop("cov_report", True)
+    if not kwargs.pop("coverage", False):
+        yield
+        return
+    with NamedTemporaryFile("w") as rcfile:
+        rcfile.write(
+            """
 [run]
 branch = True
 concurrency = multiprocessing
 """
-
-    if coverage_packages:
-        pkgs = ",".join(coverage_packages)
-        click.echo(f"Coverage enabled of packages: {pkgs}")
-        coveragerc += f"source_pkgs={pkgs}"
-    else:
-        click.echo("Coverage enabled")
-
-    with tempfile.NamedTemporaryFile(mode="w") as cov_file:
-        cov_file.write(coveragerc)
-        cov_file.flush()
-        cov = coverage.Coverage(config_file=cov_file.name)
-
-    cov.start()
-    return cov
-
-
-def _stop_coverage(cov: coverage.Coverage) -> None:
-    cov.stop()
-    outdir = "./covhtml"
-    cov.html_report(directory=outdir)
-    cov.report()
-    click.echo(f"Coverage data written to {outdir}")
+        )
+        if packages:
+            packages_str = ",".join(packages)
+            rcfile.write(f"source_pkgs = {packages_str}\n")
+        rcfile.flush()
+        cov = coverage.Coverage(config_file=rcfile.name)
+        cov.start()
+        try:
+            yield
+        finally:
+            cov.stop()
+            cov.save()
+            if report:
+                outdir = "./covhtml"
+                cov.html_report(directory=outdir)
+                click.echo(f"Coverage report written to {outdir}.")
 
 
 @click.command(cls=PipetaskCommand, epilog=epilog)
 @click.pass_context
-@ctrlMpExecOpts.coverage_option()
```

Comment:
I'm a bit confused as to how coverage options can ever be enabled if we have removed them. Is it missing a:

```python
ctrlMpExecOpts.coverage_options()
```
?

Reply:
They were added back in `optionGroups.py`, by adding them to the `qgraph_options` decorator called a few lines below.

FWIW, this file organization that effectively splits the definition of script signatures across 3-4 files drives me nuts, and it may be what I'm really bothered by when I say I dislike click (compared to argparse).  But here I'm just trying to follow the existing pattern.

Reply:
We try to put the click definitions in one place and the implementation in another because it's easier to test the implementation that way.